### PR TITLE
Add leader count check in capacity balance and add trace logs.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**

### CODE REVIEW

File: server/balancer_worker.go [hunk 1]
```diff
@@ -169,10 +174,12 @@ func (bw *balancerWorker) doBalance() error {
 		}
 		if balanceOperator == nil {
 			stats.Increment("balance.select.none")
-			return nil
+			continue
```
Comment (shenli) [inline server/balancer_worker.go]: Why change this?

File: server/cache.go [hunk 2]
```diff
@@ -374,28 +388,50 @@ func (r *regionsInfo) randRegion(storeID uint64) *metapb.Region {
...
-	if s.stats.GetCapacity() == 0 {
+	if s.stats.stats.GetCapacity() == 0 {
+		return 0
+	}
+
+	return float64(s.stats.stats.GetCapacity()-s.stats.stats.GetAvailable()) / float64(s.stats.stats.GetCapacity())
+}
+
+// leaderScore is the leader peer count score of store.
+func (s *storeInfo) leaderScore(regionCount int) float64 {
```
Comment (shenli) [inline server/cache.go]: regionCount is not in storeInfo?

File: server/cluster_worker.go [hunk 3]
```diff
@@ -52,22 +58,21 @@ func (c *raftCluster) handleRegionHeartbeat(region *metapb.Region, leader *metap
...
+	if err != nil {
+		// Do balance failed, remove it.
+		log.Errorf("do balance for region %d failed %s", regionID, err)
 		c.balancerWorker.removeBalanceOperator(regionID)
+		c.balancerWorker.removeRegionCache(regionID)
 	}
-
-	if err != nil {
-		// failed, remove it.
+	if ret {
```
Comment (shenli) [inline server/cluster_worker.go]: Rename ret to finish or succ?

File: server/cache.go [hunk 4]
```diff
@@ -340,6 +340,20 @@ func (r *regionsInfo) heartbeat(region *metapb.Region, leaderPeer *metapb.Peer)
 	return resp, nil
 }
 
+func (r *regionsInfo) leaderPeerCount(storeID uint64) int {
```
Comment (siddontang) [inline server/cache.go]: leaderRegionCount may be better.

File: server/cache.go [hunk 5]
```diff
@@ -340,6 +340,20 @@ func (r *regionsInfo) heartbeat(region *metapb.Region, leaderPeer *metapb.Peer)
...
+	r.RLock()
+	defer r.RUnlock()
+
+	return len(r.leaders.storeRegions[storeID])
```
Comment (siddontang) [inline server/cache.go]: no need to check ok if storeID is invalid.

File: server/cache.go (hunk 2 above)
Comment (qiuyesuifeng) [inline server/cache.go]: `regionCount` is only the region count in this store, not the total region count.

File: server/balancer_worker.go (hunk 1 above)
Comment (qiuyesuifeng) [inline server/balancer_worker.go]: I only want to build balancer operations more quickly,  maybe we can discuss it.
@siddontang  what is your opinion?

File: server/cache.go (hunk 4 above)
Comment (qiuyesuifeng) [inline server/cache.go]: OK

File: server/cache.go (hunk 5 above)
Comment (qiuyesuifeng) [inline server/cache.go]: Needless, if storeID is not existed in storeRegions, the result will be 0.

File: server/balancer_worker.go (hunk 1 above)
Comment (siddontang) [inline server/balancer_worker.go]: I think it is ok here, for speedup.

File: server/cluster_worker.go [hunk 6]
```diff
@@ -44,6 +44,12 @@ func (c *raftCluster) addDefaultBalanceOperator(region *metapb.Region, leader *m
 }
 
 func (c *raftCluster) handleRegionHeartbeat(region *metapb.Region, leader *metapb.Peer) (*pdpb.RegionHeartbeatResponse, error) {
+	// If the region peer count is 0, then we should not handle this.
+	if len(region.GetPeers()) == 0 {
+		log.Warnf("invalid region, zero region peer count - %v", region)
```
Comment (siddontang) [inline server/cluster_worker.go]: how can we meet this? This is a very serious bug.

Comment (qiuyesuifeng) [inline server/cluster_worker.go]: Only a defensive check and trace log, have not met yet.

Comment (siddontang) [inline server/cluster_worker.go]: Then use error instead.

**Comment by shenli** [comment]
LGTM

File: server/cache.go [hunk 7]
```diff
@@ -374,28 +388,50 @@ func (r *regionsInfo) randRegion(storeID uint64) *metapb.Region {
...
 }
 
 // usedRatio is the used capacity ratio of storage capacity.
 func (s *storeInfo) usedRatio() float64 {
-	if s.stats.GetCapacity() == 0 {
+	if s.stats.stats.GetCapacity() == 0 {
+		return 0
+	}
+
+	return float64(s.stats.stats.GetCapacity()-s.stats.stats.GetAvailable()) / float64(s.stats.stats.GetCapacity())
```
Comment (siddontang) [inline server/cache.go]: seem no format the code here? why "-" has no space around.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**